Service exploration

mmccool opened the pull request:
Resolves #324 

Adds a new exploration mechanism which is basically the same as "self-description", but allows hosting of TDs that can describe Things other than itself, and allows hosting a TD on something which is not a Thing (e.g. a plain old web service), although there are some requirements for support of HEAD, etc. that need to be observed.

NOTE: Security assertions still need to be cleaned up.  For now I kept them (reworded a bit to be consistent with hosting) but added an ed note that the current ones are redundant/inconsistent and we should just refer to the other work we've done to specify these more carefully.

See also PR #330; *only one of #330 and this PR should be merged, they are two alternatives for solving the same problem*.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/mmccool/wot-discovery/pull/325.html" title="Last updated on Jun 1, 2022, 4:14 PM UTC (177d3e9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-discovery/325/412212d...mmccool:177d3e9.html" title="Last updated on Jun 1, 2022, 4:14 PM UTC (177d3e9)">Diff</a>

farshidtz commented on index.html:
```diff
@@ -788,6 +849,99 @@ <h4>`ThingLink`</h4>
+                <dd>
+                        <p class="ednote" title="Redundant">
+                        The following two assertions on TLS and authentication
+                        are redundant and not as specfic as those provided elsewhere.
```
```suggestion
                        are redundant and not as specific as those provided elsewhere.
```

farshidtz commented on index.html:
```diff
@@ -788,6 +849,99 @@ <h4>`ThingLink`</h4>
+                        </p>
+                    <p>
+                        <span class="rfc2119-assertion" id="service-http-secure"> 
+                            An HTTP server providing a <a>TD</a> 
+                            SHOULD be using TLS.
```
TLS for?
```suggestion
                            An HTTP server providing a <a>TD</a> SHOULD use TLS when serving requests.
```

mmccool replied:
I will accept, although this whole assertion will be going away; I want to replace it with a reference to assertions we already have under Security and Privacy considerations.

mmccool commented:
I cleaned up the PR to rebase it against main, so hopefully diffs make more sense now.  I also updated the figure and the text to use "Thing Description Service" instead of just "Service" ("Server" would have required a lot more changes...).  I also update the section header and the figure to use "Thing Description Directory" rather than just "Directory" to be consistent.

While I was at it I added some arrowheads to the optional backlinks in the figure.

mmccool commented:
I also moved the `overview-svc.{png,svg}` images to `overview.{png,svg}` since if this is merged, the new figure will not be an alternative, but the actual figure.

farshidtz commented on index.html:
```diff
@@ -156,9 +156,10 @@
-           and authorization.  This document normatively defines two Exploration
-           services, one for WoT Thing self-description with a single 
-           WoT Thing Description and a searchable WoT Thing Description Directory
+           and authorization.  This document normatively defines three Exploration
+           services: for distributing a single WoT Thing Description from a regular
+           web service; for WoT Thing self-description providing a single 
+           WoT Thing Description; and a searchable WoT Thing Description Directory
```
```suggestion
           and authorization.  This document normatively defines three Exploration
           services: for distributing a single WoT Thing Description from a regular
           web server; for WoT Thing self-description providing a single 
           WoT Thing Description; and a searchable WoT Thing Description Directory
```
Or "as a regular web service". I will use web server in the next suggestions.

mmccool replied:
I did change "Thing Description Service" to "Thing Description Server" generally, but may have missed a few...

farshidtz commented on index.html:
```diff
@@ -223,8 +224,9 @@ <h1>Introduction</h1>
            entities acting on their behalf.  The WoT Discovery mechanism however also supports
-           self-description by individual Things and one issue we address is how to distinguish
-           between these two approaches.
+           self-description by individual Things and distribution of single TDs from regular
+           web services.  One issue we address is how to distinguish
+           between these approaches.
```
```suggestion
           self-description by individual Things and distribution of single TDs from regular
           web servers.  One issue we address is how to distinguish
           between these approaches.
```

farshidtz commented on index.html:
```diff
@@ -286,7 +288,7 @@ <h1>Architecture</h1>
          In this document we specify details on several Introduction
          mechanisms, suitable for different use cases, including both local and non-local
-         scenarios, but Introductions can in fact be provided any mechanism that can return a URL.
+         scenarios, but Introductions can in fact be provided by any mechanism that can return a URL.
```
```suggestion
         scenarios, but Introduction mechanisms can in fact be provided by any mechanism that can return a URL.
```

farshidtz commented on index.html:
```diff
@@ -303,6 +305,8 @@ <h1>Architecture</h1>
         <p>
 	       Each URL provided by the Introduction phase always points at an 
          Exploration service endpoint that will return a single Thing Description.
+	       In the simplest case this is the Thing Description of a Thing 
+	       describing an IoT endpoint device.  
```
```suggestion
	       In the simplest case this is the Thing Description of a Thing 
	       describing an IoT device.  
```

farshidtz commented on index.html:
```diff
@@ -984,9 +988,101 @@ <h3>Security Bootstrapping</h3>
         </section>
+
+        <section id="exploration-service" class="normative">
+            <h2>Thing Description Service</h2>
+            <p>
+             Any web service that can be referenced by a URL and return a TD with
+             appropriate authentication and access controls can be used as an
+             exploration mechanism.  
+             This service does not need to be a Thing.
```
```suggestion
            <h2>Thing Description Server</h2>
            <p>
             Any web service that can return a TD with
             appropriate authentication and access controls can be used as an
             exploration mechanism.  
             The server providing this service does not need to be a Thing.
```

farshidtz commented on index.html:
```diff
@@ -984,9 +988,101 @@ <h3>Security Bootstrapping</h3>
+             and referenced by its URL.  
+            </p>
+            <p>
+             However, a service distributing a TD 
+             using the following protocols 
+             is subject to the following constraints:
```
```suggestion
             However, a server distributing a TD 
             using the following protocols 
             is subject to the following constraints:
```

mmccool replied:
outdated now

farshidtz reviewed (changes requested):
Thanks for the updates.
I think there is a confusing use of web server and web service. In particular a web server is the component that provides a web service. The service is providing a TD. 

https://stackoverflow.com/questions/2205788/is-there-a-difference-between-a-web-server-and-a-web-service

Unfortunately that means the diagrams should be updated too. IMO, the name of component should be "TD Server".

![](https://raw.githubusercontent.com/w3c/wot-discovery/eaf94f94249f34f3091eb1c522d69ebb8381e652/images/overview.png)

Why do we need the green boundary with a different label around the exploration services? I'd just remove those green dotted lines and labels.

mmccool commented:
Also modified the figure to take out CoAP and multicast from CoRE-RD.  Technically you can use HTTP and unicast for CoRE-RD.  Also updated the class diagram.

Finally, a created another PR (PR #330, as a branch off this one) that merges self-description and TD server into one exploration mechanism, since a self-describing Thing is just a Thing that is also a TD server.  Unfortunately I branched that before merging Farshid's suggestions so if we go with PR #330 instead of this one (merging both does not make sense, they are alternative ways to address the same issue...) then we should do similar cleanups in the result.

BTW in either case we need to add "Thing Description Server" (short form, TD Server) to the Terminology definitions in Architecture.

mmccool commented:
merged alternative
